### PR TITLE
Initial commit to add support for restoring and permanently deleting SPs

### PR DESCRIPTION
**Pull request by sureshja**
Support for:

- Listing deleted SPs
- Getting a deleted SP
- Restoring a deleted SP
- Permanently deleting a deleted SP

### CODE REVIEW

File: api-reference/beta/api/directory-deleteditems-delete.md [hunk 1]
```diff
@@ -20,7 +20,8 @@ Currently, deleted items functionality is only supported for the [application](.
 ## Permissions
 One of the following permissions is required to call this API. To learn more, including how to choose permissions, see [Permissions](/graph/permissions-reference).
 
-For applications:
+### For applications and servicePrincipals:
```
Comment (jasondemorrow) [inline api-reference/beta/api/directory-deleteditems-delete.md:23]: Are we sure SPs should require the same set of permissions? Should SP delete/restore require ServicePrincipal.ReadWrite.All instead?

Comment (jasondemorrow) [reply api-reference/beta/api/directory-deleteditems-delete.md:23]: Also, Directory.ReadWrite.All below is a superset of all other perms and shouldn't be listed here.

File: api-reference/beta/api/directory-deleteditems-list.md [hunk 2]
```diff
@@ -21,7 +21,7 @@ Currently, deleted items functionality is only supported for the [application](.
 
 One of the following permissions is required to call this API. To learn more, including how to choose permissions, see [Permissions](/graph/permissions-reference).
 
-### For applications:
+### For applications and servicePrincipals:
```
Comment (jasondemorrow) [inline api-reference/beta/api/directory-deleteditems-list.md:24]: Same comments as above

File: api-reference/beta/api/directory-deleteditems-restore.md [hunk 3]
```diff
@@ -22,7 +22,7 @@ A recently deleted item will remain available for up to 30 days. After 30 days,
 ## Permissions
 One of the following permissions is required to call this API. To learn more, including how to choose permissions, see [Permissions](/graph/permissions-reference).
 
-### For applications:
+### For applications and servicePrincipals:
```
Comment (jasondemorrow) [inline api-reference/beta/api/directory-deleteditems-restore.md:25]: Same

File: api-reference/beta/api/directory-deleteditems-delete.md (hunk 1 above)
Comment (sureshja) [reply api-reference/beta/api/directory-deleteditems-delete.md:23]: Re: ServicePrincipal.ReadWrite.All - We do not have more granular permission for SP yet for MSGraph. 

Re: Directory.ReadWrite.All - We list most restrictive to least restrictive permissions. So I am going to leave it as is.

File: api-reference/beta/api/directory-deleteditems-get.md [hunk 4]
```diff
@@ -20,7 +20,7 @@ Currently, deleted items functionality is only supported for the [application](.
 ## Permissions
 One of the following permissions is required to call this API. To learn more, including how to choose permissions, see [Permissions](/graph/permissions-reference).
 
-### For applications:
+### For applications and servicePrincipals:
```
[the remark this answers is not in the document]
Comment (sureshja) [reply api-reference/beta/api/directory-deleteditems-get.md:23]: Resolving based on the response provided above.

File: api-reference/beta/api/directory-deleteditems-list.md (hunk 2 above)
Comment (sureshja) [reply api-reference/beta/api/directory-deleteditems-list.md:24]: Resolving based on the response provided above.

File: api-reference/beta/api/directory-deleteditems-restore.md (hunk 3 above)
Comment (sureshja) [reply api-reference/beta/api/directory-deleteditems-restore.md:25]: Resolving based on the response provided above.

File: api-reference/beta/api/directory-deleteditems-list.md [hunk 5]
```diff
@@ -49,6 +49,7 @@ One of the following permissions is required to call this API. To learn more, in
 <!-- { "blockType": "ignored" } -->
 ```http 
 GET /directory/deleteditems/microsoft.graph.application
+GET /directory/deleteditems/microsoft.graph.serviceprincipal
```
[the remark this answers is not in the document]
Comment (sureshja) [reply api-reference/beta/api/directory-deleteditems-list.md:52]: I addressed this in my recent check in.

File: api-reference/beta/api/directory-deleteditems-delete.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (sureshja) [reply api-reference/beta/api/directory-deleteditems-delete.md:23]: I addressed this in my recent check in.

File: api-reference/beta/api/directory-deleteditems-get.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sureshja) [reply api-reference/beta/api/directory-deleteditems-get.md:23]: I addressed this in my recent check in.

File: api-reference/beta/api/directory-deleteditems-list.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sureshja) [reply api-reference/beta/api/directory-deleteditems-list.md:24]: I addressed this in my recent check in.

File: api-reference/beta/api/directory-deleteditems-restore.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sureshja) [reply api-reference/beta/api/directory-deleteditems-restore.md:25]: I addressed this in my recent check in.

**Comment by FaithOmbongi** [comment]
Gentle ping @sureshja ...bumping this up in your queue. I believe I heard this capability you're introducing being requested during the customer onsite - application track. Kindly look into getting this PR merged soon.

**Review by FaithOmbongi — APPROVED** [review]

**Comment by FaithOmbongi** [comment]
Heads up @sureshja , I've pushed a few commits to get this PR to a healthy state. I see the functionality is still not live, so I won't add the ready to merge label. When the feature is live, please replace the **do not merge** with the **ready to merge** label.

**Comment by FaithOmbongi** [comment]
Gentle ping, @sureshja. I see this feature is already live. Could you get this PR in a healthy state for merge?

**Pull request merged**